Feature#10 main checker view

JorgeRodz opened the pull request:
# Description
What did you implemented/Why did you implemented this:
- Implement the check-in and check-out functionality in order to register an assistant
- Implement 'flash_messages' in order to display messages to the user
- Add 'live clock time' using JS vanilla

 Examples:
 - Check-in a user and store employee_id, store_id, check_in (datetime), check_out
 - Display live-clock in order to show the user at what time it is
 - Display messages in order to inform the user the form submit state

  ## Screenshots
![attendances](https://user-images.githubusercontent.com/13999498/190266002-a10910ff-ee1b-4e36-86e3-d9da5f17a849.png)

![check-in_registration](https://user-images.githubusercontent.com/13999498/190266023-4038c011-b552-4ace-b558-f64eeacefde9.gif)
![check-out_registration](https://user-images.githubusercontent.com/13999498/190266029-50a3236a-d409-4a55-a0e4-f9252b3a177a.gif)

GustavoHiguera committed:
Add javascript, html and css to show a live clock on the home page

    Co-authored-by: dagp-ucol <dgentner@ucol.mx>
    Co-authored-by: JorgeRodz <jarf4321@gmail.com>
    Co-authored-by: CalixtoLopez <calixto.lopez.santos@gmail.com>

kalixtosh committed:
Change styles employee show
Co-authored-by: dagp-ucol <dgentner@ucol.mx>
Co-authored-by: JorgeRodz <jarf4321@gmail.com>
 Co-authored-by: GustavoHiguera <ghiguera@ucol.mx>

GustavoHiguera committed:
Modify attendance views for a better presentation

Co-authored-by: dagp-ucol <dgentner@ucol.mx>
Co-authored-by: CalixtoLopez <calixto.lopez.santos@gmail.com>

Commit:
Add Check-in validation

Co-authored-by: GustavoHiguera <ghiguera@ucol.mx>
Co-authored-by: JorgeRodz <jarf4321@gmail.com>
Co-authored-by: CalixtoLopez <calixto.lopez.santos@gmail.com>

JorgeRodz committed:
Add - 'flash message' when check-in is successfully

Co-authored-by: GustavoHiguera <ghiguera@ucol.mx>
Co-authored-by: dgentner <dgentner@ucol.mx>
Co-authored-by: CalixtoLopez <calixto.lopez.santos@gmail.com>

OswaldoPineda commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -2,6 +2,20 @@
 // They will automatically be included in application.css.
 // You can use Sass (SCSS) here: https://sass-lang.com/
 
+@import url('https://fonts.googleapis.com/css?family=Orbitron');
+
+$base-duration: 500ms;
+
+// Colors
+$color-1: slategray;
+$color-2: #3498db;
+$color-3: #bfbfbf;
+
+// Breakpoints
+$sm: new-breakpoint(min-width 320px);
+$med: new-breakpoint(min-width 768px);
+$lg: new-breakpoint(min-width 1024px);
+
```
it would be better to create a new file only for variables like color or fonts, pls refactor this

OswaldoPineda commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -33,3 +48,65 @@
+      flex-direction: row;
+      align-items: center;
+      justify-content: center;
+      //width: 85px;
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -33,3 +48,65 @@
+    border-radius: 10px;
+
+    font-family: 'Orbitron', sans-serif;
+    //font-family: 'Menlo', monospace;
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -33,3 +48,65 @@
+      text-shadow: 0 0 15px rgba($color-3, 0.4);
+      line-height: 1.75;
+
+      // background-color: white;
+      // border: 1px solid red;
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -33,3 +48,65 @@
+      }
+    }
+  }
+}
```
add a final line this warning is because this file doesn't have a line at the end

JuanequeX replied:
i agree with @OswaldoPineda

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
+  end
+
+  def create
+    # puts "params: #{params}"
+
+    # Where return an array of objects, so we save the first one and thus the only one object because we have only allow to have just check_in per day
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
+    # Where return an array of objects, so we save the first one and thus the only one object because we have only allow to have just check_in per day
+    already_check_in = Attendance.where(employee_id: params[:attendance][:employee_id], check_in: Date.today.all_day).first
+
+    # byebug
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
+
+    # --------------- Check-in ---------------
+    if params[:commit] == "Check-in"
+      # attendance new instance
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
+    # byebug
+
+    # --------------- Check-in ---------------
+    if params[:commit] == "Check-in"
```
you could refactor this by doing something like
```

return if params[:commit] != "Check-in"
# resto of your code
```

on this way you removed the nesting level of the rest of this method

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
+      @attendance = Attendance.new(params.require(:attendance).permit(:employee_id,:store_id))
+      @employee = Employee.find_by(id: @attendance.employee_id)
+
+      # puts @employee.private_number
+      # puts params[:attendance][:private_number]
+
+      # byebug
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
+
+      # byebug
+
+      if @employee == nil || @employee.working == false
+        flash[:notice] = "Employee not found"
+        render 'pages/home'
+        return
+      end
+
+      if @employee.private_number != params[:attendance][:private_number].to_i
+        flash[:notice] = "Wrong private number"
+        render 'pages/home'
+        return
+      end
+
+      if already_check_in
+        flash[:notice] = "You already checked in at -> '#{already_check_in.check_in.to_s(:long)}'"
+        render 'pages/home'
+        return
```
your method has a lot of lines pls refactor this maybe in small methods and call them here

PudinDeveloper replied:
We tried but we could not, so we made an issue to work on it later.

https://github.com/BC-MAY-21-ROR/kata-06-el-reloj-checador-k6-ca-reilero-04/issues/20

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
 
-    def edit
-      @attendance = Attendance.find(params[:id])
+      # byebug
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
+        return
+      end
+
+      # attendance new instance
+      # byebug
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
-      # else
-      #   render 'edit'
+    # --------------- Check-out ---------------
+    if params[:commit] == "Check-out"
```
same as the above comment related to the nesting level and refactor

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
-      # @store = Store.find(params[:id])
-      # @store.destroy
+  def edit
+    # @attendance = Attendance.find(params[:id])
```
delete unnecessary comments

OswaldoPineda commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
+      render 'pages/home'
+      return
     end
+    # flash[:notice] = "update was reach"
+    # redirect_to root_path
+    # render 'pages/home'
+  end
+
+  def destroy
+    # @store = Store.find(params[:id])
+    # @store.destroy
+  end
 end
```
delete unnecessary comments

OswaldoPineda reviewed (changes requested):
let me know if you have any doubt but pls attend all the comments

JuanequeX commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -33,3 +48,65 @@
 // ----- Digital-clock -----
+
+.alarm-clock {
+  position: relative;
+  padding: 10px;
+  border-radius: 10px;
+  background-color: #232323;
```
sort this block alphabetically and change your background color for a variable.

JuanequeX commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -33,3 +48,65 @@
+  background-color: #232323;
+
+  .date {
+    position: absolute;
+    font-weight: bold;
+    bottom: 15px;
+    left: 50%;
+    color: #80b5ff;
+    font-size: 12px;
+    text-transform: uppercase;
+    transform: translateX(-50%);
+    z-index: 9;
```
same here, sort alphabetically and change your color for a variable.

JuanequeX commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -33,3 +48,65 @@
+  }
+
+  .time {
+    display: flex;
+    align-items: center;
+    justify-content: center;
+
+    position: relative;
+    padding: 0px 45px 15px 40px;
+    background-color: #262626;
+    border-radius: 10px;
```
sort it and change this color for a variable

JuanequeX commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -33,3 +48,65 @@
+    font-size: 52px;
+
+    span {
+      position: relative;
+      display: flex;
+      flex-direction: row;
+      align-items: center;
+      justify-content: center;
```
sort this block too

JuanequeX commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
+  def new
+    @attendance = Attendance.new
+
+    #hol
```
delete this comment

JuanequeX commented on time_checker/app/views/employees/new.html.erb:
```diff
@@ -1,51 +1,53 @@
+    </p>
+
+  <% end %>
+</div>
```
Add a blank line here

JuanequeX commented on time_checker/app/views/pages/_clock.html.erb:
```diff
@@ -0,0 +1,15 @@
+    <span class="colon">:</span>
+    <span class="seconds"></span>
+  </div>
+</div>
```
Add a blank line here

JuanequeX commented on time_checker/app/views/stores/new.html.erb:
```diff
@@ -1,30 +1,32 @@
+      <td><%= link_to 'Cancel and return to store index', stores_path, class: 'btn btn-primary' %></td>
+    </p>
+  <% end %>
+</div>
```
Add an blank line here!

juperameza commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,103 @@
 
-    def index
-      @attendances = Attendance.all
-    end
```
Try to delete trailing blank spaces

juperameza replied:
I use this vs extension https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces

PudinDeveloper replied:
## Your extension is not working haha.

![image](https://user-images.githubusercontent.com/73558208/190282187-990b5c7e-a92a-4c0a-85da-13d2674fc14d.png)

juperameza replied:
Fixed I added a bracket. Lol

PudinDeveloper replied:
XD Thanks :D

JuanequeX commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -2,20 +2,31 @@
 // They will automatically be included in application.css.
 // You can use Sass (SCSS) here: https://sass-lang.com/
 
+@import "./variables/colors.scss";
```
if they are use the pattern 7-1 in scss, are not necessary import this file.

JuanequeX commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -2,20 +2,31 @@
 // You can use Sass (SCSS) here: https://sass-lang.com/
 
+@import "./variables/colors.scss";
+@import url('https://fonts.googleapis.com/css?family=Orbitron');
```
This import should be in a file called `fonts.scss`

JuanequeX commented on time_checker/app/assets/stylesheets/pages.scss:
```diff
@@ -2,20 +2,31 @@
+@import "./variables/colors.scss";
+@import url('https://fonts.googleapis.com/css?family=Orbitron');
+
+$base-duration: 500ms;
+
+// Breakpoints
+$sm: new-breakpoint(min-width 320px);
+$med: new-breakpoint(min-width 768px);
+$lg: new-breakpoint(min-width 1024px);
+
```
All of this variables shouldn't exist here, it's will be declared in other file, this breakpoints in `mixins.scss` file  for example

JorgeRodz committed:
Refactor - attendances controller and apply correct format on files:

- app/controllers/employees_controller.rb
- app/controllers/pages_controller.rb
- app/models/attendance.rb

JorgeRodz committed:
Add - check-in, check-out functionality:

- Add 'attendance controller' methods for check-in, check-out
- Add flash_messages

Co-authored-by: GustavoHiguera <ghiguera@ucol.mx>
Co-authored-by: dgentner <dgentner@ucol.mx>
Co-authored-by: CalixtoLopez <calixto.lopez.santos@gmail.com>

JorgeRodz committed:
Add the code that reviewers requested to change and run rubocop -a on root folder

JorgeRodz committed:
Refactor - attendances_controller and add 'today_attendances' scope to attendance model

Co-authored-by: GustavoHiguera <ghiguera@ucol.mx>
Co-authored-by: dgentner <dgentner@ucol.mx>
Co-authored-by: CalixtoLopez <calixto.lopez.santos@gmail.com>

JorgeRodz committed:
Merge branch 'refactorr-feature#10-main-checker-view' into feature#10-main-checker-view

Co-authored-by: GustavoHiguera <ghiguera@ucol.mx>
Co-authored-by: dgentner <dgentner@ucol.mx>
Co-authored-by: CalixtoLopez <calixto.lopez.santos@gmail.com>

kalixtosh committed:
Fix request change styles
GustavoHiguera <ghiguera@ucol.mx>
  Co-authored-by: dgentner <dgentner@ucol.mx>
Co-authored-by: JorgeRodz < jarf4321@gmail.com>

juperameza commented on time_checker/app/controllers/attendances_controller.rb:
```diff
@@ -1,40 +1,99 @@
+
+  def destroy
+    # @store = Store.find(params[:id])
+    # @store.destroy
```
Don't leave unused code 🙅🙅

JorgeRodz replied:
We already delete that code because we no longer use it.

juperameza commented on time_checker/app/controllers/employees_controller.rb:
```diff
@@ -1,39 +1,44 @@
+
+  def update
+    @employee = Employee.find(params[:id])
+    if @employee.update(params.require(:employee).permit(:first_name, :last_name, :email, :position, :working,
```
Here I recommend you use another method to define your strong Params since it's being used twice and repeating code
[Like this!](https://api.rubyonrails.org/classes/ActionController/StrongParameters.html)

JorgeRodz replied:
Refactor done on every controller.

JorgeRodz committed:
Refactor - stores controller, employees controller, attendances
controller and views:

- Refactor params object in all controller mention above
- Set initial variables for all controller mention above
- Delete edit view on attendances views

Co-authored-by: GustavoHiguera <ghiguera@ucol.mx>
Co-authored-by: dgentner <dgentner@ucol.mx>
Co-authored-by: CalixtoLopez <calixto.lopez.santos@gmail.com>

iMartinDav reviewed (approved):
LGTM

iMartinDav reviewed (approved):
Looks good

willcab reviewed (approved):
good job

juperameza reviewed (approved):
LGMT

OswaldoPineda reviewed (approved):
LGTM 👍🏽

JuanequeX reviewed (approved):
LGTM
